File: tests/fixtures/core-frames.ts
import type {CoreFrameCatalog,CorePointTransform} from "../../packages/contracts/src/spatial/core/frame-schema";

export const frameRef=(id:string,revision=1)=>({ref:{namespace:"frame" as const,id},revision});
export const opRef=(id:string,revision=1)=>({ref:{namespace:"transform" as const,id},revision});
const benchmark=(id:string)=>({kind:"benchmark" as const,reference:{ref:{namespace:"benchmark",id},revision:1},label:id});
export function frameFixture():CoreFrameCatalog {
  return {
    frames:[
      {ref:{namespace:"frame",id:"feet"},revision:1,label:"North/east feet, positive down",sourceCrs:null,kind:"engineering",horizontalUnit:"ft",axes:["north","east"],verticalUnit:"ft",verticalDirection:"down",vertical:benchmark("source-zero")},
      {ref:{namespace:"frame",id:"metric"},revision:1,label:"Metre ENU",sourceCrs:null,kind:"engineering",horizontalUnit:"m",axes:["east","north"],verticalUnit:"m",verticalDirection:"up",vertical:benchmark("target-zero")},
      {ref:{namespace:"frame",id:"ecef"},revision:1,label:"WGS84 ECEF",sourceCrs:"EPSG:4978",kind:"geocentric",crs:"EPSG:4978",unit:"m",axes:"X-Y-Z"},
    ],
    operations:[
      {ref:{namespace:"transform",id:"registration"},revision:1,from:frameRef("feet"),to:frameRef("metric"),provenance:"Authored numeric fixture; not a survey",accuracyMetres:null,kind:"local_rigid",rotationDegrees:90,translationMetres:[10,20],sourceDomain:{minEast:-100,minNorth:-100,maxEast:100,maxNorth:100},verticalTie:{kind:"constant_offset",offsetMetres:100}},
      {ref:{namespace:"transform",id:"world"},revision:1,from:frameRef("metric"),to:frameRef("ecef"),provenance:"Authored equatorial benchmark",accuracyMetres:null,kind:"wgs84_enu",sourceDomain:null,benchmark:benchmark("target-zero").reference,origin:{longitude:0,latitude:0,ellipsoidHeightMetres:100}},
    ],
  };
}
export function frameRequest():CorePointTransform {
  return {from:frameRef("feet"),to:frameRef("metric"),point:[10,20,5],steps:[{operation:opRef("registration"),direction:"forward"}]};
}
export interface FrameConformanceCase {id:string;catalog:unknown;request:unknown;valid:boolean;expectedPoint?:readonly number[];tolerance?:number;code?:string}
export function frameCases():FrameConformanceCase[] {
  const cases:FrameConformanceCase[]=[];
  const add=(id:string,expectedPoint:readonly number[]|string,change:(c:any,r:any)=>void=()=>{},tolerance=1e-10)=>{
    const catalog=JSON.parse(JSON.stringify(frameFixture())),request=JSON.parse(JSON.stringify(frameRequest()));change(catalog,request);
    cases.push({id,catalog,request,valid:typeof expectedPoint!=="string",...(typeof expectedPoint==="string"?{code:expectedPoint}:{expectedPoint,tolerance})});
  };
  add("feet-permuted-axes-rotation-vertical-tie",[6.952,26.096,98.476]);
  add("inverse-registration",[10,20,5],(_c,r)=>{r.from=frameRef("metric");r.to=frameRef("feet");r.point=[6.952,26.096,98.476];r.steps[0].direction="inverse";});
  add("horizontal-only-needs-no-vertical-tie",[6.952,26.096],(c,r)=>{c.frames[0].vertical={kind:"unknown",reason:"Not provided"};c.operations[0].verticalTie={kind:"unavailable",reason:"No level tie"};r.point=[10,20];});
  add("same-frame-no-op",[10,20,5],(_c,r)=>{r.to=frameRef("feet");r.steps=[];});
  add("millimetres-to-metres",[1,2,3],(c,r)=>{c.frames[0].horizontalUnit="mm";c.frames[0].verticalUnit="mm";c.frames[0].verticalDirection="up";c.frames[0].axes=["east","north"];c.frames[0].vertical=c.frames[1].vertical;c.operations[0].rotationDegrees=0;c.operations[0].translationMetres=[0,0];c.operations[0].verticalTie={kind:"same_reference"};r.point=[1000,2000,3000];});
  add("legacy-survey-foot-is-not-international-foot",[3.048006096012192,6.096012192024384],(c,r)=>{c.frames[0].horizontalUnit="us_ft";c.frames[0].axes=["east","north"];c.operations[0].rotationDegrees=0;c.operations[0].translationMetres=[0,0];r.point=[10,20];});
  add("west-south-axis-signs",[-3.048,-6.096],(c,r)=>{c.frames[0].axes=["west","south"];c.operations[0].rotationDegrees=0;c.operations[0].translationMetres=[0,0];r.point=[10,20];});
  add("negative-vertical-tie",[6.952,26.096,-11.524],c=>{c.operations[0].verticalTie.offsetMetres=-10;});
  const world=(c:any,r:any)=>{r.from=frameRef("metric");r.to=frameRef("ecef");r.point=[1,2,3];r.steps=[{operation:opRef("world"),direction:"forward"}];};
  add("equatorial-wgs84-basis",[6378240,1,2],world,1e-8);
  add("equatorial-east-90-basis",[-1,6378140,2],(c,r)=>{world(c,r);c.operations[1].origin={longitude:90,latitude:0,ellipsoidHeightMetres:0};},1e-8);
  add("polar-wgs84-basis",[-2,1,6356755.314245179],(c,r)=>{world(c,r);c.operations[1].origin={longitude:0,latitude:90,ellipsoidHeightMetres:0};},1e-8);
  add("inverse-world-basis",[1,2,3],(c,r)=>{world(c,r);r.from=frameRef("ecef");r.to=frameRef("metric");r.point=[6378240,1,2];r.steps[0].direction="inverse";},1e-8);
  add("two-step-independent-world-answer",[6378335.476,6.952,26.096],(_c,r)=>{r.to=frameRef("ecef");r.steps.push({operation:opRef("world"),direction:"forward"});},1e-8);
  add("stale-source-frame","STALE_REFERENCE",c=>{c.operations[0].from.revision=2;});
  add("missing-target-frame","MISSING_REFERENCE",c=>{c.operations[0].to=frameRef("missing");});
  add("duplicate-operation","DUPLICATE_RECORD",c=>c.operations.push(c.operations[0]));
  add("duplicate-frame","DUPLICATE_RECORD",c=>c.frames.push(c.frames[0]));
  add("axes-are-not-two-horizontal-directions","FRAME_AXES",c=>{c.frames[0].axes=["east","west"];});
  add("reversed-source-domain","FRAME_BOUNDS",c=>{c.operations[0].sourceDomain.minEast=101;});
  add("unknown-vertical-cannot-be-fixed-by-offset","VERTICAL_UNRESOLVED",c=>{c.frames[0].vertical={kind:"unknown",reason:"Absent"};});
  add("surface-depth-cannot-be-fixed-by-offset","VERTICAL_UNRESOLVED",c=>{c.frames[0].vertical={kind:"surface_relative",surface:{ref:{namespace:"surface",id:"ground"},revision:1},label:"Depth below terrain"};});
  add("unavailable-vertical-blocks-3d","VERTICAL_UNRESOLVED",c=>{c.operations[0].verticalTie={kind:"unavailable",reason:"No tie"};});
  add("different-benchmarks-not-same-reference","VERTICAL_MISMATCH",c=>{c.operations[0].verticalTie={kind:"same_reference"};});
  add("one-benchmark-cannot-have-two-zeros","VERTICAL_MISMATCH",c=>{c.frames[0].vertical=c.frames[1].vertical;});
  add("ellipsoidal-origin-not-default-zero","INVALID_CONTRACT",c=>{delete c.operations[1].origin.ellipsoidHeightMetres;});
  add("latitude-out-of-bounds","INVALID_CONTRACT",c=>{c.operations[1].origin.latitude=91;});
  add("world-tie-to-wrong-benchmark","VERTICAL_MISMATCH",c=>{c.operations[1].benchmark=benchmark("source-zero").reference;});
  add("world-requires-explicit-z","POINT_DIMENSION",(c,r)=>{world(c,r);r.point=[1,2];});
  add("stale-operation","STALE_REFERENCE",(_c,r)=>{r.steps[0].operation.revision=2;});
  add("disjoint-path","TRANSFORM_PATH",(_c,r)=>{r.steps=[{operation:opRef("world"),direction:"forward"}];});
  add("cyclic-path","TRANSFORM_CYCLE",(_c,r)=>{r.to=frameRef("feet");r.steps.push({operation:opRef("registration"),direction:"inverse"});});
  add("empty-path-cannot-change-frame","TRANSFORM_PATH",(_c,r)=>{r.steps=[];});
  add("outside-forward-domain","OUTSIDE_DOMAIN",(_c,r)=>{r.point=[1000,1000];});
  add("outside-inverse-domain","OUTSIDE_DOMAIN",(_c,r)=>{r.from=frameRef("metric");r.to=frameRef("feet");r.point=[1000,1000];r.steps[0].direction="inverse";});
  add("unqualified-geodetic-profile","TRANSFORM_PROFILE",c=>{const {rotationDegrees,translationMetres,sourceDomain,verticalTie,...base}=c.operations[0];c.operations[0]={...base,kind:"unsupported",profile:"grid-shift",reason:"No qualified grid operation"};});
  add("source-precision-is-not-invented",[6.952,26.096],(_c,r)=>{r.point=[10,20];});
  return cases;
}

File: services/geo/geo/core_frames.py
"""Bounded engineering registrations and explicit WGS84 ENU/ECEF placement.

The generated schema is the structural authority. This module executes only the
declared profiles; source coordinate records are never modified in place.
"""
import math
from .core_contract import bundled_schema, fail, index_records, parse_core, require_revision, version_key


def _east_west(axis):
    return axis in ("east", "west")


def _sign(axis):
    return -1 if axis in ("west", "south", "down") else 1


def _known_vertical(frame):
    return frame["kind"] != "geocentric" and frame["vertical"]["kind"] in ("benchmark", "datum")


def _same_vertical(a, b):
    return (_known_vertical(a) and _known_vertical(b)
            and a["vertical"]["kind"] == b["vertical"]["kind"]
            and version_key(a["vertical"]["reference"]) == version_key(b["vertical"]["reference"]))


def validate_core_frame_catalog(value):
    catalog = parse_core("frame-catalog", value)
    frames = index_records(catalog["frames"], "frame")
    index_records(catalog["operations"], "transform")
    for frame in catalog["frames"]:
        if frame["kind"] in ("engineering", "projected") and _east_west(frame["axes"][0]) == _east_west(frame["axes"][1]):
            fail("FRAME_AXES", "A planar frame requires independent east/west and north/south axes")
    for operation in catalog["operations"]:
        source = require_revision(frames, operation["from"], "source frame")
        target = require_revision(frames, operation["to"], "target frame")
        if version_key(operation["from"]) == version_key(operation["to"]):
            fail("TRANSFORM_SELF", "Use an empty path for an unchanged frame")
        if operation["kind"] == "unsupported":
            continue
        if source["kind"] != "engineering":
            fail("FRAME_PROFILE", "Engineering source required")
        domain = operation["sourceDomain"]
        if domain and (domain["minEast"] > domain["maxEast"] or domain["minNorth"] > domain["maxNorth"]):
            fail("FRAME_BOUNDS", "Transform source bounds are reversed")
        if operation["kind"] == "local_rigid":
            if target["kind"] != "engineering":
                fail("FRAME_PROFILE", "Engineering target required")
            tie = operation["verticalTie"]
            if tie["kind"] != "unavailable":
                if not _known_vertical(source) or not _known_vertical(target):
                    fail("VERTICAL_UNRESOLVED", "Constant ties cannot establish unknown/surface-relative height")
                if tie["kind"] == "same_reference" and not _same_vertical(source, target):
                    fail("VERTICAL_MISMATCH", "Declared vertical references differ")
                if tie["kind"] == "constant_offset" and _same_vertical(source, target) and tie["offsetMetres"] != 0:
                    fail("VERTICAL_MISMATCH", "One exact reference cannot have two elevation zeros")
        else:
            if target["kind"] != "geocentric":
                fail("FRAME_PROFILE", "WGS84 ENU requires WGS84 ECEF target")
            if source["vertical"]["kind"] != "benchmark" or version_key(source["vertical"]["reference"]) != version_key(operation["benchmark"]):
                fail("VERTICAL_MISMATCH", "Ellipsoidal origin must tie the exact local benchmark")
    return catalog


def _horizontal(frame, point, factors):
    factor = factors[frame["horizontalUnit"]]
    first = point[0] * factor * _sign(frame["axes"][0])
    second = point[1] * factor * _sign(frame["axes"][1])
    return (first, second) if _east_west(frame["axes"][0]) else (second, first)


def _up(frame, value, factors):
    return value * factors[frame["verticalUnit"]] * _sign(frame["verticalDirection"])


def _encode(frame, east, north, up, factors):
    factor = factors[frame["horizontalUnit"]]
    result = [(east if _east_west(frame["axes"][i]) else north) * _sign(frame["axes"][i]) / factor for i in range(2)]
    if up is not None:
        result.append(up * _sign(frame["verticalDirection"]) / factors[frame["verticalUnit"]])
    return result


def _domain(operation, east, north):
    bounds = operation["sourceDomain"]
    if bounds and (east < bounds["minEast"] or east > bounds["maxEast"] or north < bounds["minNorth"] or north > bounds["maxNorth"]):
        fail("OUTSIDE_DOMAIN", "Point is outside the declared source-domain bounds")


def _local(operation, source, target, point, inverse, factors):
    input_frame, output_frame = (target, source) if inverse else (source, target)
    east, north = _horizontal(input_frame, point, factors)
    dx, dy = operation["translationMetres"]
    angle = operation["rotationDegrees"] * math.pi / 180
    c, s = math.cos(angle), math.sin(angle)
    if inverse:
        output_east = (east - dx) * c + (north - dy) * s
        output_north = -(east - dx) * s + (north - dy) * c
        _domain(operation, output_east, output_north)
    else:
        _domain(operation, east, north)
        output_east, output_north = east * c - north * s + dx, east * s + north * c + dy
    up = None
    if len(point) == 3:
        tie = operation["verticalTie"]
        if tie["kind"] == "unavailable":
            fail("VERTICAL_UNRESOLVED", "No qualified vertical tie supplied")
        offset = tie["offsetMetres"] if tie["kind"] == "constant_offset" else 0
        up = _up(input_frame, point[2], factors) + (-offset if inverse else offset)
    return _encode(output_frame, output_east, output_north, up, factors)


def _enu(operation, source, point, inverse, policy):
    if len(point) != 3:
        fail("POINT_DIMENSION", "World placement requires explicit three-dimensional coordinates")
    origin = operation["origin"]
    lon, lat = origin["longitude"] * math.pi / 180, origin["latitude"] * math.pi / 180
    sl, cl, sp, cp = math.sin(lon), math.cos(lon), math.sin(lat), math.cos(lat)
    flattening = 1 / policy["wgs84"]["inverseFlattening"]
    e2 = flattening * (2 - flattening)
    n = policy["wgs84"]["semiMajorMetres"] / math.sqrt(1 - e2 * sp * sp)
    h = origin["ellipsoidHeightMetres"]
    translation = [(n + h) * cp * cl, (n + h) * cp * sl, (n * (1 - e2) + h) * sp]
    east_axis, north_axis, up_axis = [-sl, cl, 0], [-sp * cl, -sp * sl, cp], [cp * cl, cp * sl, sp]
    factors = policy["lengthMetres"]
    if inverse:
        delta = [point[i] - translation[i] for i in range(3)]
        east, north, up = [sum(delta[i] * axis[i] for i in range(3)) for axis in (east_axis, north_axis, up_axis)]
        _domain(operation, east, north)
        return _encode(source, east, north, up, factors)
    east, north = _horizontal(source, point, factors)
    up = _up(source, point[2], factors)
    _domain(operation, east, north)
    return [east * east_axis[i] + north * north_axis[i] + up * up_axis[i] + translation[i] for i in range(3)]


def transform_core_point(catalog_value, request_value):
    catalog = validate_core_frame_catalog(catalog_value)
    request = parse_core("point-transform", request_value)
    policy = bundled_schema("frame-catalog")[0]["x-ulpin-frame-policy"]
    frames, operations = index_records(catalog["frames"], "frame"), index_records(catalog["operations"], "transform")
    source = require_revision(frames, request["from"], "request source")
    target = require_revision(frames, request["to"], "request target")
    if any(frame["kind"] == "geocentric" for frame in (source, target)) and len(request["point"]) != 3:
        fail("POINT_DIMENSION", "ECEF points require three coordinates")
    if len(request["point"]) == 3 and any(frame["kind"] != "geocentric" and not _known_vertical(frame) for frame in (source, target)):
        fail("VERTICAL_UNRESOLVED", "Unknown or surface-relative height does not establish a 3D tie")
    point, cursor, accuracies = request["point"], request["from"], []
    visited = {version_key(cursor)}
    for step in request["steps"]:
        operation = require_revision(operations, step["operation"], "requested transform")
        inverse = step["direction"] == "inverse"
        input_frame, output_frame = (operation["to"], operation["from"]) if inverse else (operation["from"], operation["to"])
        if version_key(input_frame) != version_key(cursor):
            fail("TRANSFORM_PATH", "Operation path is not frame/revision continuous")
        if version_key(output_frame) in visited:
            fail("TRANSFORM_CYCLE", "Operation path repeats a frame")
        if operation["kind"] == "unsupported":
            fail("TRANSFORM_PROFILE", "Transform profile has no qualified executor")
        source = require_revision(frames, operation["from"], "operation source")
        target = require_revision(frames, operation["to"], "operation target")
        if operation["kind"] == "local_rigid":
            point = _local(operation, source, target, point, inverse, policy["lengthMetres"])
        else:
            point = _enu(operation, source, point, inverse, policy)
        if not all(math.isfinite(value) for value in point):
            fail("NON_FINITE_RESULT", "Coordinate arithmetic overflowed the supported profile")
        visited.add(version_key(output_frame))
        cursor = output_frame
        accuracies.append(operation["accuracyMetres"])
    if version_key(cursor) != version_key(request["to"]):
        fail("TRANSFORM_PATH", "Operation path does not reach the requested frame revision")
    return {"point": point, "frame": request["to"], "steps": request["steps"], "declaredAccuraciesMetres": accuracies}
